functions/src/cron-jobs/specialists.ts

```diff
@@ -0,0 +1,111 @@
+  const { AIRTABLE_API_KEY, AIRTABLE_BASE } = getSecrets();
+  const allSkills: SkillRecord[] = [];
+
+  return new Promise((resolve, reject) => {
+    const airtable = getAirtable(AIRTABLE_API_KEY, AIRTABLE_BASE);
+    airtable(BASE.SKILLS)
+      .select({
+        view: BASE_VIEWS.ALL,
+        fields: [SKILL_FIELDS.NAME, SKILL_FIELDS.INTERMEDIATE, SKILL_FIELDS.EXPERT],
+      })
+      .eachPage(
+        function page(records: any, fetchNextPage: any) {
+          allSkills.push(
+            ...records.map((record: any) => {
+              const intermediate = record.get(SKILL_FIELDS.INTERMEDIATE) || [];
+              const experts = record.get(SKILL_FIELDS.EXPERT) || [];
+              const specialistsAmount = experts.length + intermediate.length;
+
+              const skill: SkillRecord = {
+                label: record.get(SKILL_FIELDS.NAME),
+                specialistsAmount,
+              };
+              return skill;
+            })
+          );
+
+          fetchNextPage();
+        },
+        function done(err: any) {
+          if (err) {
+            functions.logger.error(`Failed to get Airtable data: ${err}`);
+            return reject({});
+          }
+          const skillsReduced = allSkills.reduce((result, skill) => {
+            result[skill.label] = skill.specialistsAmount;
+            return result;
+          }, {} as Record<string, number>);
+
+          return resolve(skillsReduced);
+        }
+      );
+  });
+};
+
+const getContentfulData = async (): Promise<Entry[]> => {
+  return new Promise(async (resolve, reject) => {
+    const { CONTENTFUL_TOKEN, CONTENTFUL_SPACE_ID, CONTENTFUL_ENVIRONMENT } = getSecrets();
+    try {
+      const client = getContentfulClient(CONTENTFUL_TOKEN);
+      const contentfulConfig = { space: CONTENTFUL_SPACE_ID, environment: CONTENTFUL_ENVIRONMENT };
+      const environment = await getEnvironment(client, contentfulConfig);
+      const entries = await environment.getEntries({ content_type: 'entry', limit: 700 });
+      resolve(entries.items);
+    } catch (err) {
+      functions.logger.error(`Failed to get Contentful data: ${err}`);
+      reject([]);
+    }
+  });
+};
+
+export const updateSpecialistsAmount = functions
+  .runWith({ secrets: [...AIRTABLE_SECRET_KEYS, ...CONTENTFUL_SECRET_KEYS, ...URL_SECRET_KEYS] })
+  .pubsub.schedule('0 9 1,15 * *') // runs at 1st and 15th day of every month 9:00
+  .timeZone(DEFAULT_TIMEZONE)
+  .onRun(async () => {
+    const allSkills = await getAirtableData();
+    const contentfulEntries = await getContentfulData();
+    const entries = contentfulEntries.map((entry) => {
+      const label = entry.fields.label?.['en-US'];
+      const currentAmount = entry.fields.experts?.['en-US'];
+      return {
+        label,
+        newAmount: allSkills[label],
+        currentAmount: Number(currentAmount ?? 0),
+        entry,
+      };
+    });
+    entries.forEach((technology, i) => {
+      setTimeout(() => {
```

Comment:
what is this `setTimeout` for? is it to workaround API limits? maybe you can use bulk publish instead https://www.contentful.com/developers/docs/references/content-management-api/#/reference/bulk-actions/publish-bulk-action?

Reply:
This method you’ve sent is not made for bulk updates and won’t work in this case.

As I was digging more, I found out that `forEach` is not made for async operations, and my previous approach was bad because it was not waiting for a promise to finish, so I’ve decided to change it into for of loop - which works great, and by using it I’m not hitting the rate limit.

Take a look at this article at contentful blog https://www.contentful.com/blog/bulk-operations-programamatic-content-management-how-dfds-lost-its-head-part-two/ - it’s the same approach as mine with using `publish` and `update` methods on entries.